tools/shadowsim/src/shadowsim/shadow_shell.py

```diff
@@ -0,0 +1,179 @@
+
+        # Run specified command, merging stderr to stdout
+        exec {arg_str} 2>&1
+        """
+    )
+
+    config = scfg.Config(
+        general=scfg.General(
+            # Experimentally, this is about the highest we can go before
+            # the time spent fast forwarding empty time gets noticeable.
+            stop_time="1000000h",
+            log_level="warning",
+            heartbeat_interval=f"{2**63}ns",
+        ),
+        network=scfg.Network(graph=scfg.Graph(type="1_gbit_switch")),
+        hosts={
+            "host": scfg.Host(
+                network_node_id=0,
+                processes=[
+                    scfg.Process(
+                        path="bash",
+                        args=[
+                            "-c",
+                            wrapper_script,
+                        ],
+                    )
+                ],
+            )
+        },
+    )
+    config_path = tmpdir.joinpath("shadow.yaml")
+    config_path.write_text(yaml.safe_dump(config))
+
+    data_dir = tmpdir.joinpath("shadow.data")
+    shadow_args = [
+        f"--data-directory={data_dir}",
+        "--progress=false",
+        str(config_path),
+    ]
+
+    shadow_log_path = tmpdir.joinpath("shadow.log")
+    with shadow_log_path.open("w") as shadow_log:
+        shadow_ps = subprocess.Popen(
+            ["shadow"] + shadow_args,
+            stdout=shadow_log,
+            stderr=subprocess.PIPE,
+        )
+        shadow_stdout = None
+        shadow_exited = False
+        while True:
+            # Try opening the simulated process's stdout if we haven't
+            # successfully done so yet.
+            if shadow_stdout is None:
+                try:
+                    shadow_stdout = open(
+                        str(data_dir.joinpath("hosts/host/bash.1000.stdout")), "rb"
+                    )
+                except FileNotFoundError:
+                    # Not created yet, presumably
+                    pass
+
+            # Pump data from sim stdout to our stdout
+            data = None
+            if shadow_stdout is not None:
+                data = shadow_stdout.read()
+            if data:
+                stdout.write(data)
+                stdout.flush()
+
+            if not data and shadow_exited:
+                # Done
+                break
+
+            if not data:
+                # Wait a bit for data to become ready, or shadow to exit.  We
+                # might be able to do something with inotify to be notified
+                # about the stdout file being created or having data ready, but
+                # it seems more complex than it's worth.
+                try:
+                    shadow_ps.wait(1)
+                    # Mark as exited, but loop again in case more output has
+                    # been written.
+                    shadow_exited = True
+                except subprocess.TimeoutExpired:
+                    pass
```

Comment:
Do we want to sleep a tiny bit here inside the except block `if shadow_stdout is None` to prevent completely annihilating a core in case it takes a bit for Shadow to start up the requested process?

Reply:
Oops. I think I totally misread that; the `wait(1)` call just got done sleeping a tiny bit, right?

Reply:
Right. Reworked the comments here a bit.